primer/src/Primer/App.hs

```diff
@@ -894,31 +931,47 @@ applyProgAction prog = \case
           newName <- LocalName <$> freshName (freeVars e)
           binds' <- maybe (throwError $ IndexOutOfRange index) pure $ insertAt index (Bind m' newName) binds
           pure $ CaseBranch vc binds' e
-  DeleteConField tdName vcName index -> editModuleCross (qualifiedModule tdName) prog $ \(m, ms) -> do
-    m' <-
-      alterTypeDef
-        ( \td -> do
-            checkTypeNotInUse tdName td $ m : ms
-            traverseOf
-              #astTypeDefConstructors
-              ( maybe (throwError $ ConNotFound vcName) pure
-                  <=< findAndAdjustA
-                    ((== vcName) . valConName)
-                    (traverseOf #valConArgs $ maybe (throwError $ IndexOutOfRange index) pure . map fst . deleteAt index)
-              )
-              td
+  DeleteConField tdName vcName index -> editModuleOfCrossType (Just tdName) prog $ \(m, ms) defName def -> do
+    def' <-
+      traverseOf
+        #astTypeDefConstructors
+        ( maybe (throwError $ ConNotFound vcName) pure
+            <=< findAndAdjustA
+              ((== vcName) . valConName)
+              (traverseOf #valConArgs $ map fst . deleteIndex)
         )
-        tdName
-        m
+        def
+    ms' <-
+      ((m & over #moduleTypes (Map.insert defName $ TypeDefAST def')) : ms)
+        & traverseOf
+          (traversed % #moduleDefs % traversed % #_DefAST % #astDefExpr)
+          ( transformNamedCaseBranch
+              tdName
+              vcName
+              ( \_ (CaseBranch vc binds e) -> do
+                  (binds', Bind _ var) <- deleteIndex binds
+                  CaseBranch vc binds'
+                    <$> ( e & transformM \case
+                            Var _ v | v == LocalVarRef var -> emptyHole
+                            e' -> pure e'
+                        )
+              )
```

Comment:
I'm suspicious of this inner "v ~> ?" transformation, in the presence of shadowing. What would happen when deleting the parameter in `case e of C x -> MkPair x (\x -> x)`, where the last `x` does not refer to the pattern variable? Perhaps `traverse` over `_freeVars` would be more appropriate?

Reply:
Good catch, that's a pretty silly mistake from me! I've now replaced with a `traverseOf _freeTmVars`. I'd have hoped that our property tests might catch this sort of thing, but I guess there's nothing biasing the generators towards creating shadowing scenarios.

Reply:
> However, I don't know what test would actually catch this, as it would still be type-correct, just delete more that expected.

Oh, good point.